### PR TITLE
Issue/#1292 Localize authentication_failed messages

### PR DESCRIPTION
**Pull request by Askaholic**
Related #1292.

Add localization for some messages that are currently hard coded strings in the lobby server.

Managed to to get these working with my server PR https://github.com/FAForever/server/pull/505. I plan on adding more messages in a separate PR after I get more comfortable with how all the pieces of the code work together.

### CODE REVIEW

File: src/main/java/com/faforever/client/remote/domain/AuthenticationFailedMessage.java
Comment (1-alex98) [inline src/main/java/com/faforever/client/remote/domain/AuthenticationFailedMessage.java]: java doc please

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java
Comment (1-alex98) [inline src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java:486]: So this ends in a lot of lines if we do this for every mesage. What I would expect would be a unique identifying string/number for the message. Which in return can be found in the messages of the client maybe with a prefix For Example:
Server message has a field messageId= 147
-> i18n.get("serverMessage.147", message.getParameters());
-> serverMessage.147 = Login failed, you are banned for {0} days
Translating hence can be done in one line

**Review by 1-alex98 — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java
Comment (Askaholic) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java:486]: Ok I see. I'll have to think about the best way to do this so it still makes sense for the protocol. The only thing is it doesn't handle config parameters like the website url which don't need to be sent by the server.

Comment (Askaholic) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java:486]: I don't think I would use numbers, but I might do something like `i18n.get(String.format("authenticationMessage.%s", message.getContext()), /* parameters here */)`

Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java:486]: Just saying the here presented solution seems to be a bad idea. Also what we definitely need is a fallback string that is sent if the server message is not found in message.properties

Comment (Askaholic) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java:486]: Ok. I've made it more general where it uses the value of the message identifier as the localization key. So for instance localization of a policy server error for "fraudulent" response would use key `login.error.policy.fraudulent`.

If the given key is not found, it will try to fall back by popping off the last identifier in the chain until either it finds a key, or it hits `login.error` which returns a very generic error message. So if for instance the key `login.error.policy.fraudulent` didn't exist, it would try  `login.error.policy` which does exist, and returns a generic error about a policy violation.

Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java:486]: sounds better

Comment (1-alex98) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java:486]: ok but still this is specific code for "login",  so we need to copy this for everything else right. Tho I understand that if u wanna add specific args from the clientProperties thats hard to realize. So if u want to do it this way that's fine with me.

**Review by 1-alex98 — APPROVED** [review]

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java
Comment (Askaholic) [reply src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java:486]: Well when I do the next message type I can refactor the bits I need.

**Comment by 1-alex98** [comment]
Still need to test this

**Comment by Askaholic** [comment]
I have tested it using the test server (and the PR I made for that). I tried to write some tests, but I don't think I was going about it the right way. JUnit 4 doesn't have very good support for testing that a function throws a specific exception, and I'm thinking that's probably not the best way to do it anyways. Suggestions on how to write the tests are welcome

**Comment by 1-alex98** [comment]
I meant test in run and see if it breaks

**Comment by 1-alex98** [comment]
Just test the positive cases then and u can still catch the Exceptiona and see if it is what u expect

**Comment by 1-alex98** [comment]
@Askaholic is this deployed server sided?

**Comment by Askaholic** [comment]
No, I was going to wait for this client PR to be merged first since it will continue to work with the old server messages.

**Comment by 1-alex98** [comment]
ok so I merge this?

**Comment by 1-alex98** [comment]
next time make a branch in this repo. I always need to add your fork remote otherwises

**Comment by 1-alex98** [comment]
grafik
I won't merge this if that's the message we will get. We should fallback to the or email or password is wrong message.

**Comment by Askaholic** [comment]
How did you produce that? It's supposed to prefer any message sent by the server over the generic error message:
```Java
return messageText != null ? messageText: i18n.get("login.error");
```

**Comment by 1-alex98** [comment]
I checked out your branch and entered wrong credidentials

**Review by 1-alex98 — CHANGES_REQUESTED** [review]
chnages requested see above

**Comment by 1-alex98** [comment]
So shell I try again?

**Comment by 1-alex98** [comment]
lemme se

**Comment by Askaholic** [comment]
What is the status of this PR? Have you tested it @axel1200?

**Comment by 1-alex98** [comment]
@Askaholic u realise that if we switch to api being the one that grants login that gets obsolete

**Comment by Askaholic** [comment]
The client will still need to authenticate with the server. And not all users are going to choose to use an API token. We still will support both methods of login.

**Comment by 1-alex98** [comment]
Yes but the primary login will be the api if it fails then the error message of the api will be shown

**Comment by Askaholic** [comment]
In response to your previous comment, you will still want to localize those error messages from the API.

I think I have a better way of localizing server messages in general. How about we add an `i18n_key` field to the existing `notice` message? That will require a lot less rewriting and be much more flexible in general. We can still send English text as a fallback too.

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java
Comment (1-alex98) [inline src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java:513]: What does this mean. Please define a proper fallback then...

File: src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java
Comment (1-alex98) [inline src/main/java/com/faforever/client/remote/FafServerAccessorImpl.java:514]: i18n.get(...) right?

**Review by 1-alex98 — CHANGES_REQUESTED** [review]

**Comment by 1-alex98** [comment]
Has conflicts needs to be rebased

**Comment by Sheikah45** [comment]
What is the status of this pr?

**Comment by Sheikah45** [comment]
Closing as it is stale

**Pull request closed without merging**
